### PR TITLE
Update React to v17

### LINKED ISSUE #6938: Use React 17
## Task Description

See #5793 for details.

To learn more about React 17:

**React 17 blog post**: https://reactjs.org/blog/2020/10/20/react-v17.html
**Breaking changes**: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#other-breaking-changes
**New JSX transform**: https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

To learn more about React 17:

**React 17 blog post**: https://reactjs.org/blog/2020/10/20/react-v17.html
**Breaking changes**: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#other-breaking-changes
**New JSX transform**: https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

Side effect: This also avoids some misleading `SharedArrayBuffer` warnings that were fixed in React 17.0.2, see https://github.com/facebook/react/blob/7c4e6aae3ef4e77596125f30b801c84800e949a9/CHANGELOG.md#react-dom

## Relevant Technical Choices

* Removes unneeded imports of `React` using the provided codemod (`npx react-codemod update-react-imports`).  
  This is done in preparation of using the new JSX transform, since the new JSX transform doesn’t require React to be in scope.  
  **Note:** These imports are already unneeded today, and basically dead code. It doesn’t hurt to keep them, but they can be removed safely.
* Also cleans up code in lots of places to get rid of console warnings (DOM validation warnings, incorrect prop types
* Mocks `@wordpress/element` (currently on React 16.x) in tests to map to React 17.  
  Avoids issues with mismatching React versions used for code and tests.  
  These can be removed once `@wordpress/element` upgrades too (https://github.com/WordPress/gutenberg/pull/26847)

## To-do

* Use new JSX transform (in follow-up PR)  
  Turns out this requires some changes to the Karma config and how we spy on `React.createElement`, as this spy needs to be changed to work against `react/jsx-runtime`.

## User-facing changes

N/A

## Testing Instructions

N/A

---

Fixes #6938

### CODE REVIEW

**Comment by spacedmonkey** [comment]
Can you explain why you removed all the imports of `React`? What does that do?

**Comment by swissspidy** [comment]
> Can you explain why you removed all the imports of `React`? What does that do?

It's basically dead code. Updated the PR description accordingly.

Note that the PR is still a draft because I haven't yet figured out why Moveable doesn't work like it should.

**Comment by spacedmonkey** [comment]
>It's basically dead code. 

Sorry, I just jumped the gun reviewing it. I only looked at it for about 2 minutes, just to get my head around it. I was wonder, should we remove this react import now or is this a change that is required for the 17 upgrade. Seems like description now answer question, so I am glad I asked.

**Comment by swissspidy** [comment]
Now blocked by https://github.com/reach/router/pull/436 

```
$ npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: react@17.0.1
npm ERR! node_modules/react
npm ERR!   react@"^17.0.1" from the root project
npm ERR!   peer react@">=16.3.0" from @emotion/core@10.1.1
npm ERR!   node_modules/@emotion/core
npm ERR!     peer @emotion/core@"^10.0.27" from @emotion/primitives-core@10.0.27
npm ERR!     node_modules/@emotion/primitives-core
npm ERR!       @emotion/primitives-core@"10.0.27" from @emotion/native@10.0.27
npm ERR!       node_modules/@wordpress/components/node_modules/@emotion/native
npm ERR!         @emotion/native@"^10.0.22" from @wordpress/components@12.0.5
npm ERR!         node_modules/@wordpress/components
npm ERR!     peer @emotion/core@"^10.0.27" from @emotion/styled@10.0.27
npm ERR!     node_modules/@emotion/styled
npm ERR!       @emotion/styled@"^10.0.23" from @storybook/theming@6.1.16
npm ERR!       node_modules/@storybook/theming
npm ERR!         @storybook/theming@"6.1.16" from @storybook/addon-a11y@6.1.16
npm ERR!         node_modules/@storybook/addon-a11y
npm ERR!         11 more (@storybook/addon-actions, ...)
npm ERR!       1 more (@wordpress/components)
npm ERR!     7 more (@emotion/styled-base, @storybook/theming, ...)
npm ERR!   64 more (@emotion/primitives-core, @emotion/styled, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@"15.x || 16.x || 16.4.0-alpha.0911da3" from @reach/router@1.3.4
npm ERR! node_modules/@reach/router
npm ERR!   @reach/router@"^1.3.3" from @storybook/api@6.1.16
npm ERR!   node_modules/@storybook/api
npm ERR!     @storybook/api@"6.1.16" from @storybook/addon-a11y@6.1.16
npm ERR!     node_modules/@storybook/addon-a11y
npm ERR!       dev @storybook/addon-a11y@"^6.1.14" from the root project
npm ERR!     9 more (@storybook/addon-actions, ...)
npm ERR!   @reach/router@"^1.3.3" from @storybook/router@6.1.16
npm ERR!   node_modules/@storybook/router
npm ERR!     @storybook/router@"6.1.16" from @storybook/addon-storysource@6.1.16
npm ERR!     node_modules/@storybook/addon-storysource
npm ERR!       dev @storybook/addon-storysource@"^6.1.14" from the root project
npm ERR!     4 more (@storybook/addons, @storybook/api, @storybook/core, @storybook/ui)
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

**Comment by miina** [comment]
@googlebot I consent.

**Comment by barklund** [comment]
@googlebot I consent.

File: assets/src/dashboard/app/views/editorSettings/adManagement/googleAdManager/test/googleAdManager.js
Comment (spacedmonkey) [inline assets/src/dashboard/app/views/editorSettings/adManagement/googleAdManager/test/googleAdManager.js:45]: What is this change?

Comment (swissspidy) [reply assets/src/dashboard/app/views/editorSettings/adManagement/googleAdManager/test/googleAdManager.js:45]: A bugfix to address prop type warnings.

The `GoogleAdManagerSettings` component expects a `slotId` prop, not `adManagerSlotId`.

File: assets/src/dashboard/karma/fixture.js
Comment (spacedmonkey) [inline assets/src/dashboard/karma/fixture.js:20]: What does this do?

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/common/innerElement.js:210]: Why is this change needed?

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/media/common/mediaElement.js:164]: Can we leave the comment here?

File: assets/src/dashboard/karma/fixture.js
Comment (swissspidy) [reply assets/src/dashboard/karma/fixture.js:20]: See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#import_an_entire_modules_contents

Whereas before it used the default export (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#importing_defaults)

As per the [documentation](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html):

> The default imports will keep working in React 17, but in the longer term we encourage moving away from them.

So we are moving away from them.

File: package.json
Comment (spacedmonkey) [inline package.json]: Why is this removed?

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/common/mediaElement.js:164]: It‘s not correct anymore because it‘s now always a promise

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:210]: The reference could be null

File: package.json
Comment (swissspidy) [reply package.json]: Because it‘s not needed anymore because of the new JSX transform

**Comment by spacedmonkey** [comment]
There seems to be karma tests failing and a js lint error.

**Comment by spacedmonkey** [comment]
I think we should make sure that https://github.com/google/web-stories-wp/pull/6900 is merged before merging this.

**Comment by swissspidy** [comment]
> I think we should make sure that #6900 is merged before merging this.

I don't see why. There are no real dependencies between these two. Whatever is ready first could go in.

**Review by spacedmonkey — APPROVED** [review]
Changes look good to me.

**Pull request merged**
